### DISCUSSION TITLE
JWT authentication for REST client

### DISCUSSION THREAD

**Discussion by IvanPuntev** [quarkusio/quarkus#40036] [Q&A]
For jwt authentication in my rest client interface I'm currently using custom class that extends the ReactiveClientHeadersFactory and manually building the access token in it. What is the proper way in quarkus 3 to do that? Is it possible to be as simple as the OIDC - you just add annotation @RegisterProvider(OidcClientRequestReactiveFilter.class) to the rest client and add some config properties?

quarkus.oidc-client.auth-server-url=
quarkus.oidc-client.client-id=
quarkus.oidc-client.credentials.secret=
quarkus.oidc-client.grant.type=
quarkus.oidc-client.discovery-enabled=
quarkus.oidc-client.token-path=
quarkus.oidc-client.grant-options.password.username=
quarkus.oidc-client.grant-options.password.password=

[a comment by an automated account is not shown]

**Comment by sberyozkin** [comment]
@IvanPuntev Right, that should do it, the filter will manage the token acquisition and refresh and will set it as the `Authorization: Bearer token`.
You can use `@OidcClientFilter` annotation instead of registering the filter directly

**Reply by IvanPuntev** [reply]
@sberyozkin Is there any example project? I tried configuring the client using these properties but it does not work. It throws exception io.quarkus.oidc.client.runtime.DisabledOidcClientException: 'Default' client configuration is not initialized

quarkus.oidc-client.auth-server-url=
quarkus.oidc-client.credentials.jwt.secret=
quarkus.oidc-client.credentials.jwt.issuer=
quarkus.oidc-client.credentials.jwt.subject=
quarkus.oidc-client.credentials.jwt.audience=
quarkus.oidc-client.credentials.jwt.key-file=
quarkus.oidc-client.grant.type=jwt
quarkus.oidc-client.token-path=
quarkus.oidc-client.client-id=

**Reply by sberyozkin** [reply]
@IvanPuntev See https://github.com/quarkusio/quarkus-quickstarts/pull/1402, this only adds `@OidcClientFilter`

**Reply by IvanPuntev** [reply]
@sberyozkin I've added the annotation but I'm more interested in the proper configuration properties when the grant type is jwt.
I have secret, issuer, subject, audience and private key to provide.
The example shows the config for grant type password.

**Reply by sberyozkin** [reply]
@IvanPuntev Authentication options are listed here: https://quarkus.io/guides/security-openid-connect-client-reference#oidc-client-authentication

As far as standard JWT authentication options are concerned, it is either client_secret_jwt or private_key_jwt, so you don't need to provide both the secret and private key, and other properties you have mentioned should be set automatically as JWT claims, but if needed you can override them, https://quarkus.io/guides/security-openid-connect-client-reference#additional-jwt-authentication-options, but you don't need to do it by default

**Reply by IvanPuntev** [reply]
@sberyozkin Can the private key pem contents be provided as a string instead of providing a file to the property quarkus.oidc-client.credentials.jwt.key-file= ? If not possible then can you give me a best practice how and where to store this pem file. The application is built using docker and then deployed to google cloud. The sensitive information is stored in gcp secrets as strings.

**Reply by sberyozkin** [reply]
@IvanPuntev  JWT private key PEM representations can not be specified directly as strings in the configuration, but makes sense to support it so that one can refer to them as `${my_gcp_private_key_value}`

**Reply by IvanPuntev** [reply]
Can I open a PR to add the raw string support?
Will it be better to add a separate property like  quarkus.oidc-client.credentials.jwt.key= ?

**Reply by sberyozkin** [reply]
@IvanPuntev 

>Can I open a PR to add the raw string support?

it would be great

>Will it be better to add a separate property like quarkus.oidc-client.credentials.jwt.key= ?

Indeed, adding a `key` property to the JWT group would do, there should be a method in KeyUtils which will correctly read this String, which I believe must be a Base64 representation of the private key, and it should take property over the file and keystore properties. Please give it a try and let me know how it goes

**Reply by IvanPuntev** [reply]
Should I create a feature branch from main or some other branch? I'm currently using version 3.8.3. Also should I open the PR for merging in main or some other development branch, etc.?

**Reply by sberyozkin** [reply]
@IvanPuntev Please fork the main repository and create the branch off the main branch in that fork and indeed, open a PR (may be in a draft PR for us to tune a few things) against the main branch, also see https://github.com/quarkusio/quarkus/blob/main/CONTRIBUTING.md

**Reply by sberyozkin** [reply]
@IvanPuntev I've opened #40078 to track it further, thanks

**Reply by sberyozkin** [reply]
@IvanPuntev Hopefully it will be a simple PR, please make the new property check first here, https://github.com/quarkusio/quarkus/blob/main/extensions/oidc-common/runtime/src/main/java/io/quarkus/oidc/common/runtime/OidcCommonUtils.java#L332, and if it is set, convert it to Key with this KeyUtils call: try https://github.com/smallrye/smallrye-jwt/blob/main/implementation/jwt-build/src/main/java/io/smallrye/jwt/build/impl/JwtSignatureImpl.java#L262

**Reply by IvanPuntev** [reply]
Yea, I already did some code investigation and made those changes.

**Reply by IvanPuntev** [reply]
@sberyozkin This is the pull request https://github.com/quarkusio/quarkus/pull/40086. Tell me if something else needs to be added.

**Reply by sberyozkin** [reply]
@IvanPuntev Sure, I've commented at it, many thanks for this contribution
